### PR TITLE
controller: wait for synced old replica sets on Recreate

### LINKED ISSUE #27362: Recreate deployments dont wait for pod termination
The scenario is a single replica with an attached EBS. 

The create seems to happen right after the delete not waiting for the pod to be terminated. This results in the two kubelets fighting over the EBS which generally results in AWS becoming unhappy meaning the EBS has to be force detached.

From the docs `All existing Pods are killed before new ones are created when .spec.strategy.type==Recreate` I would have thought deployments would wait until the pod is actually killed.

Is this a bug or am i misinterpreting the docs?

### PR DESCRIPTION
**Pull request by 0xmichalis**
Partially fixes https://github.com/kubernetes/kubernetes/issues/27362

Any other work on it should be handled in the replica set level (and/or kubelet if it's required)

@kubernetes/deployment PTAL

### CODE REVIEW

File: pkg/client/conditions.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (kargakis) [inline pkg/client/conditions.go]: Not sure if this is the right place for clientset conditions @krousey @caesarxuchao @smarterclayton

Comment (lavalamp) [inline pkg/client/conditions.go]: This is an unfortunate place, since it won't work with all of our various clients...

Comment (smarterclayton) [inline pkg/client/conditions.go]: Yeah, I think this belongs scoped.

**Comment by pwittrock** [comment]
@janetkuo Do you want to take a look at well?
@kargakis I plan to take a look at this next week after 1.3 calms down

**Comment by kargakis** [comment]
No rush on this

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -544,6 +545,10 @@ func (dc *DeploymentController) syncRecreateDeployment(deployment *extensions.De
 		return dc.updateDeploymentStatus(allRSs, newRS, deployment)
 	}
 
+	if err := dc.waitForSyncedReplicaSets(deployment); err != nil {
```
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Can we just pass `oldRSs` in?

File: pkg/controller/deployment/deployment_controller.go [hunk 3]
```diff
@@ -573,6 +578,23 @@ func (dc *DeploymentController) syncRecreateDeployment(deployment *extensions.De
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+func (dc *DeploymentController) waitForSyncedReplicaSets(deployment *extensions.Deployment) error {
```
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: Suggest name this to something like `waitForReplicaSetsInactive` and add some comments

File: pkg/controller/deployment/deployment_controller.go (hunk 2 above)
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Actually yes we can, will update

File: pkg/controller/deployment/deployment_controller.go (hunk 3 above)
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Updated to `waitForInactiveReplicaSets`

**Pull request merged**